pkg/generators/extension.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generators
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/golang/glog"
+	"k8s.io/gengo/types"
+)
+
+// Extension tag to openapi extension
+var tagToExtension = map[string]string{
+	"patchMergeKey": "x-kubernetes-patch-merge-key",
+	"patchStrategy": "x-kubernetes-patch-strategy",
+	"listType":      "x-kubernetes-list-type",
+	"listTypeKey":   "x-kubernetes-list-type-keys",
+}
+
+// Enum values per extension
+var allowedExtensionValues = map[string][]string{
+	"x-kubernetes-patch-strategy": []string{"merge", "retainKeys"},
+	"x-kubernetes-list-type":      []string{"atomic", "set", "map"},
+}
+
+// TODO: Create interface for an extension validation function, and create
+// validation functions for different extensions.
+type Extension struct {
+	name   string   // Example: x-kubernetes-list-type
+	values []string // Example: atomic
+}
+
+func (e Extension) validateAllowedValues() error {
+	// allowedValues not set means no restrictions on values.
+	allowedValues, exists := allowedExtensionValues[e.name]
+	if !exists {
+		return nil
+	}
+	// For each extension value, validate that it is allowed.
+	for _, value := range e.values {
+		if !valueInArray(value, allowedValues) {
+			return fmt.Errorf("%s: value (%s) not allowed", e.name, value)
+		}
+	}
+	return nil
+}
+
+func valueInArray(a string, array []string) bool {
```

Comment:
I don't know if we're allowed to import it, but the sets.Strings type does nearly all of these helper things.

Reply:
We probably don't want to make this repository dependent on Kubernetes.

Reply:
Hm, we can actually get it from here: https://godoc.org/k8s.io/gengo/examples/set-gen/sets

Any issue with grabbing it from the examples directory?

Reply:
The scope of this PR is already too large. I'll address this in another PR.

Reply:
I agree that if we can find any library to have these helpers, it is better but have reservation on using an example repo for this.